ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/FailureMessage.java

```diff
@@ -35,6 +40,29 @@ public static PubsubMessage of(Object caller, String payload, Throwable e) {
     return FailureMessage.of(caller, payload.getBytes(StandardCharsets.UTF_8), e);
   }
 
+  /**
+   * Return a PubsubMessage wrapping a SponsoredInteraction with attributes describing the error.
+   * TODO Investigate refactoring to not import SponsoredInteraction into this higher-level module.
+   */
+  public static PubsubMessage of(Object caller, SponsoredInteraction interaction, Throwable e) {
+    Map<String, String> attributes = new HashMap<String, String>();
+    attributes.put(Attribute.DOCUMENT_NAMESPACE,
+        Optional.ofNullable(interaction.getOriginalNamespace()).orElse("missing_namespace"));
+    attributes.put(Attribute.DOCUMENT_TYPE,
+        Optional.ofNullable(interaction.getOriginalDocType()).orElse("missing_doctype"));
+    attributes.put(Attribute.SUBMISSION_TIMESTAMP, Optional //
+        .ofNullable(interaction.getSubmissionTimestamp()) //
+        .orElseGet(() -> Time.epochMicrosToTimestamp(new Instant().getMillis() * 1000)));
+    Map<String, String> additionalAttributes = interaction.getAdditionalAttributes();
+    if (additionalAttributes != null) {
+      attributes.putAll(interaction.getAdditionalAttributes());
+    }
```

Comment:
I'd recommend putting additionalAttributes into the map first, so that the explicit fields of `SponsoredInteraction` will always take precedence if we happen to duplicate a field.